worlds/yugioh06/rules.py

```diff
@@ -0,0 +1,878 @@
+        if loc.name in location_rules:
+            add_rule(loc, location_rules[loc.name])
+        if loc.name in access_rules:
+            add_rule(multiworld.get_entrance(loc.name, player), access_rules[loc.name])
+
+
+def only_light(state, player):
+    return state.count_from_list_exclusive([
+        "Dunames Dark Witch",
+        "X-Head Cannon",
+        "Homunculus the Alchemic Being",
+        "Hysteric Fairy",
+        "Ninja Grandmaster Sasuke"], player) >= 2 \
+           and state.count_from_list_exclusive([
+        "Chaos Command Magician",
+        "Cybernetic Magician",
+        "Kaiser Glider",
+        "The Agent of Judgment - Saturn",
+        "Zaborg the Thunder Monarch",
+        "Cyber Dragon"], player) >= 1 \
+           and state.count_from_list_exclusive([
+        "D.D. Warrior Lady",
+        "Mystic Swordsman LV2",
+        "Y-Dragon Head",
+        "Z-Metal Tank",
+    ], player) >= 2 and state.has("Shining Angel", player)
+
+
+def only_dark(state, player):
+    return state.count_from_list_exclusive([
+        "Dark Elf",
+        "Archfiend Soldier",
+        "Mad Dog of Darkness",
+        "Vorse Raider",
+        "Skilled Dark Magician",
+        "Skull Descovery Knight",
+        "Mechanicalchaser",
+        "Dark Blade",
+        "Gil Garth",
+        "La Jinn the Mystical Genie of the Lamp",
+        "Opticlops",
+        "Zure, Knight of Dark World",
+        "Brron, Mad King of Dark World",
+        "D.D. Survivor",
+        "Exarion Universe",
+        "Kycoo the Ghost Destroyer",
+        "Regenerating Mummy"
+    ], player) >= 2 \
+           and state.count_from_list_exclusive([
+        "Summoned Skull",
+        "Skull Archfiend of Lightning",
+        "The End of Anubis",
+        "Dark Ruler Ha Des",
+        "Beast of Talwar",
+        "Inferno Hammer",
+        "Jinzo",
+        "Ryu Kokki"
+    ], player) >= 1 \
+           and state.count_from_list_exclusive([
+        "Legendary Fiend",
+        "Don Zaloog",
+        "Newdoria",
+        "Sangan",
+        "Spirit Reaper",
+        "Giant Germ"
+    ], player) >= 2 and state.has("Mystic Tomato", player)
+
+
+def only_earth(state, player):
+    return state.count_from_list_exclusive([
+        "Berserk Gorilla",
+        "Gemini Elf",
+        "Insect Knight",
+        "Toon Gemini Elf",
+        "Familiar-Possessed - Aussa",
+        "Neo Bug",
+        "Blindly Loyal Goblin",
+        "Chiron the Mage",
+        "Gearfried the Iron Knight"
+    ], player) >= 2 and state.count_from_list_exclusive([
+        "Dark Driceratops",
+        "Granmarg the Rock Monarch",
+        "Hieracosphinx",
+        "Saber Beetle"
+    ], player) >= 1 and state.count_from_list_exclusive([
+        "Hyper Hammerhead",
+        "Green Gadget",
+        "Red Gadget",
+        "Yellow Gadget",
+        "Dimensional Warrior",
+        "Enraged Muka Muka",
+        "Exiled Force"
+    ], player) >= 2 and state.has("Giant Rat", player)
```

Comment:
Oh nice, you switched to `count_from_list_exclusive`.

However, if you're just comparing to a single number, you should do:

`state.has_from_list_exclusive([...], player, 2)`
instead of
`state.count_from_list_exclusive([...], player) >= 2`

It will almost always be more performant for some technical reasons. Keep in mind that the `state.has` functions mean "has at least" :)

Please make this change whereever possible